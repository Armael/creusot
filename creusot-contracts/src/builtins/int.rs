use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
