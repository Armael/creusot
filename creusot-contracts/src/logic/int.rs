use super::model::*;
use crate as creusot_contracts;
use crate::logic::*;
use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        absurd
    }
}

impl From<u8> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn from(_: u8) -> Self {
        absurd
    }
}
impl Model for u8 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        absurd
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: u64) -> Self {
        absurd
    }
}
impl Model for u64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn from(_: i64) -> Self {
        absurd
    }
}
impl Model for i64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        absurd
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        absurd
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        absurd
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
