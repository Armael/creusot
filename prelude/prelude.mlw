module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
