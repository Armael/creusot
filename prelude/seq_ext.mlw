module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end