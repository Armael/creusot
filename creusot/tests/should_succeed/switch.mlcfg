
module Switch_Option_Type
  type switch_option_type 't =
    | Switch_Option_Some_Type 't
    | Switch_Option_None_Type
    
  let function switch_option_type_Some_0 (self : switch_option_type 't) : 't = [@vc:do_not_keep_trace] [@vc:sp]
    match (self) with
      | Switch_Option_Some_Type a -> a
      | Switch_Option_None_Type -> any 't
      end
end
module Switch_Test_Interface
  use mach.int.Int
  use mach.int.UInt32
  use Switch_Option_Type
  val test [@cfg:stackify] (o : Switch_Option_Type.switch_option_type uint32) : bool
end
module Switch_Test
  use mach.int.Int
  use mach.int.UInt32
  use Switch_Option_Type
  use prelude.IntSize
  let rec cfg test [@cfg:stackify] [#"../switch.rs" 9 0 9 35] (o : Switch_Option_Type.switch_option_type uint32) : bool
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : bool;
  var o_1 : Switch_Option_Type.switch_option_type uint32;
  var _2 : isize;
  var x_3 : uint32;
  var _4 : uint32;
  {
    o_1 <- o;
    goto BB0
  }
  BB0 {
    switch (o_1)
      | Switch_Option_Type.Switch_Option_Some_Type _ -> goto BB3
      | Switch_Option_Type.Switch_Option_None_Type -> goto BB1
      end
  }
  BB1 {
    _0 <- false;
    goto BB4
  }
  BB2 {
    absurd
  }
  BB3 {
    x_3 <- Switch_Option_Type.switch_option_type_Some_0 o_1;
    _4 <- x_3;
    _0 <- ([#"../switch.rs" 11 19 11 24] _4 > (0 : uint32));
    goto BB4
  }
  BB4 {
    return _0
  }
  
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Impl0_Resolve_Interface
  type t1
  type t2
  predicate resolve (self : (t1, t2))
end
module CreusotContracts_Logic_Resolve_Impl0_Resolve
  type t1
  type t2
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface as Resolve1 with type self = t2
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface as Resolve0 with type self = t1
  predicate resolve (self : (t1, t2)) =
    Resolve0.resolve (let (a, _) = self in a) /\ Resolve1.resolve (let (_, a) = self in a)
end
module CreusotContracts_Logic_Resolve_Impl0
  type t1
  type t2
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve_Interface
  type t
  predicate resolve (self : t)
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve
  type t
  predicate resolve (self : t) =
    true
end
module CreusotContracts_Logic_Resolve_Impl2
  type t
end
module Switch_Test2_Interface
  use mach.int.Int
  use mach.int.UInt32
  use Switch_Option_Type
  val test2 [@cfg:stackify] (o : (Switch_Option_Type.switch_option_type uint32, uint32)) : uint32
end
module Switch_Test2
  use mach.int.Int
  use mach.int.UInt32
  use Switch_Option_Type
  use prelude.IntSize
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve2 with type t = uint32
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = Switch_Option_Type.switch_option_type uint32
  clone CreusotContracts_Logic_Resolve_Impl0_Resolve as Resolve0 with type t1 = Switch_Option_Type.switch_option_type uint32,
  type t2 = uint32, predicate Resolve0.resolve = Resolve1.resolve, predicate Resolve1.resolve = Resolve2.resolve
  let rec cfg test2 [@cfg:stackify] [#"../switch.rs" 16 0 16 42] (o : (Switch_Option_Type.switch_option_type uint32, uint32)) : uint32
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : uint32;
  var o_1 : (Switch_Option_Type.switch_option_type uint32, uint32);
  var _2 : isize;
  var x_3 : uint32;
  {
    o_1 <- o;
    goto BB0
  }
  BB0 {
    switch (let (a, _) = o_1 in a)
      | Switch_Option_Type.Switch_Option_Some_Type _ -> goto BB3
      | Switch_Option_Type.Switch_Option_None_Type -> goto BB1
      end
  }
  BB1 {
    _0 <- (let (_, a) = o_1 in a);
    assume { Resolve0.resolve o_1 };
    goto BB4
  }
  BB2 {
    assume { Resolve0.resolve o_1 };
    absurd
  }
  BB3 {
    x_3 <- Switch_Option_Type.switch_option_type_Some_0 (let (a, _) = o_1 in a);
    assume { Resolve0.resolve o_1 };
    _0 <- x_3;
    goto BB4
  }
  BB4 {
    return _0
  }
  
end
