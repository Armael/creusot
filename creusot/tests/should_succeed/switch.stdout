module Type
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type switch_option 't = 
    | Switch_Option_Some('t)
    | Switch_Option_None
    
  predicate drop_switch_option (drop_t : 't -> bool) (self : switch_option 't) = 
    match (self) with
      | Switch_Option_Some(a) -> drop_t a
      | Switch_Option_None -> true
      end
end
module Switch_Test
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  let rec cfg impl (o_o_1 : Type.switch_option uint32) : bool = 
  var _0 : bool;
  var o_1 : Type.switch_option uint32;
  var _2 : isize;
  var x_3 : uint32;
  var _4 : uint32;
  {
    o_1 <- o_o_1;
    goto BB0
  }
  BB0 {
    switch (o_1)
      | Type.Switch_Option_Some(_) -> goto BB1
      | Type.Switch_Option_None -> goto BB2
      | _ -> goto BB3
      end
  }
  BB1 {
    assume { drop_int _2 };
    goto BB4
  }
  BB2 {
    assume { Type.drop_switch_option drop_uint o_1 };
    assume { drop_int _2 };
    _0 <- false;
    goto BB5
  }
  BB3 {
    assume { Type.drop_switch_option drop_uint o_1 };
    assume { drop_int _2 };
    absurd
  }
  BB4 {
    assume { drop_uint x_3 };
    x_3 <- (let Type.Switch_Option_Some(a) = o_1 in a);
    assume { Type.drop_switch_option drop_uint o_1 };
    assume { drop_uint _4 };
    _4 <- x_3;
    assume { drop_uint x_3 };
    _0 <- _4 > (0 : uint32);
    goto BB5
  }
  BB5 {
    return _0
  }
  
end
module Switch_Test2
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  let rec cfg impl (o_o_1 : (Type.switch_option uint32, uint32)) : uint32 = 
  var _0 : uint32;
  var o_1 : (Type.switch_option uint32, uint32);
  var _2 : isize;
  var x_3 : uint32;
  {
    o_1 <- o_o_1;
    goto BB0
  }
  BB0 {
    switch (let (a, _) = o_1 in a)
      | Type.Switch_Option_Some(_) -> goto BB1
      | Type.Switch_Option_None -> goto BB2
      | _ -> goto BB3
      end
  }
  BB1 {
    assume { drop_int _2 };
    goto BB4
  }
  BB2 {
    assume { drop_int _2 };
    assume { drop_uint _0 };
    _0 <- (let (_, a) = o_1 in a);
    assume { (fun tup -> let (a, b) = tup in Type.drop_switch_option drop_uint a && drop_uint b) o_1 };
    goto BB5
  }
  BB3 {
    assume { (fun tup -> let (a, b) = tup in Type.drop_switch_option drop_uint a && drop_uint b) o_1 };
    assume { drop_int _2 };
    absurd
  }
  BB4 {
    assume { drop_uint x_3 };
    x_3 <- (let Type.Switch_Option_Some(a) = let (a, _) = o_1 in a in a);
    assume { (fun tup -> let (a, b) = tup in Type.drop_switch_option drop_uint a && drop_uint b) o_1 };
    assume { drop_uint _0 };
    _0 <- x_3;
    assume { drop_uint x_3 };
    goto BB5
  }
  BB5 {
    return _0
  }
  
end
module Switch_Main
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  let rec cfg impl () : () = 
  var _0 : ();
  {
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
